pkg/controller/servicednsendpoint/servicednsendpoint_controller.go

```diff
@@ -0,0 +1,464 @@
+			endpoints[i-1].Targets = sortAndRemoveDuplicateTargets(endpoints[i-1].Targets)
+
+			// Remove the duplicate endpoint
+			endpoints = append(endpoints[:i], endpoints[i+1:]...)
+			continue
+		}
+		i++
+	}
+
+	return endpoints
+}
+
+func sortAndRemoveDuplicateTargets(targets []string) []string {
+	sort.Slice(targets, func(i, j int) bool {
+		return targets[i] < targets[j]
+	})
+	for i := 1; i < len(targets); {
+		if targets[i] == targets[i-1] {
+			// Remove duplicate target
+			targets = append(targets[:i], targets[i+1:]...)
+			continue
+		}
+		i++
+	}
+	return targets
+}
+
+func getNamesForDNSObject(dnsObject *feddnsv1a1.MultiClusterServiceDNSRecord, federation, dnsSuffix string) ([]*feddnsv1a1.Endpoint, error) {
+	endpoints := []*feddnsv1a1.Endpoint{}
+
+	commonPrefix := strings.Join([]string{dnsObject.Name, dnsObject.Namespace, federation, "svc"}, ".")
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		zone := clusterDNS.Zone
+		region := clusterDNS.Region
+
+		dnsNames := []string{
+			strings.Join([]string{commonPrefix, zone, region, dnsSuffix}, "."), // zone level
+			strings.Join([]string{commonPrefix, region, dnsSuffix}, "."),       // region level, one up from zone level
+			strings.Join([]string{commonPrefix, dnsSuffix}, "."),               // global level, one up from region level
+			"", // nowhere to go up from global level
+		}
+
+		zoneTargets, regionTargets, globalTargets := getHealthyTargets(zone, region, dnsObject)
+		targets := [][]string{zoneTargets, regionTargets, globalTargets}
+
+		for i, target := range targets {
+			endpoint, err := generateEndpoint(dnsNames[i], target, dnsNames[i+1])
+			if err != nil {
+				return nil, err
+			}
+			endpoints = append(endpoints, endpoint)
+		}
+	}
+
+	return endpoints, nil
+}
+
+// getHealthyTargets returns the hostnames and/or IP addresses of healthy endpoints for the service, at a zone, region and global level (or an error)
+func getHealthyTargets(zone, region string, dnsObject *feddnsv1a1.MultiClusterServiceDNSRecord) (zoneTargets, regionTargets, globalTargets feddnsv1a1.Targets) {
+	// If federated dnsObject is deleted, return empty endpoints, so that DNS records are removed
+	if dnsObject.DeletionTimestamp != nil {
+		return zoneTargets, regionTargets, globalTargets
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		if clusterDNS.Zone == zone {
+			zoneTargets = append(zoneTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+		}
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		if clusterDNS.Region == region {
+			regionTargets = append(regionTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+		}
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		globalTargets = append(globalTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+	}
+
+	return zoneTargets, regionTargets, globalTargets
+}
+
+// Abstracting away the internet
+type NetWrapper interface {
```

Comment:
Are you expecting multiple implementations for this interface? If the answer is 'NO' "for now", probably this interface is not needed.

Reply:
This is for test purpose, you can see the unit tests for its usage.

Reply:
Probably a comment here would make this confusion go away.

Reply:
have already added the below comment in code.
```
// Abstracting away the internet for testing purposes
type NetWrapper interface {
	LookupHost(host string) (addrs []string, err error)
}
```